### ISSUE TITLE
Make prev_timestep of DDIMScheduler accessible

### ISSUE THREAD

**Issue by jenuk** [huggingface/diffusers#2165]
**Is your feature request related to a problem? Please describe.**
The DDIM-Scheduler works in theory with any subselection of the train timestep; the original authors propose for example a square schedule. However, the current implementation fixes the subselection to be a linear subselection. I would like to be able to set an arbitrary schedule at inference.

This is currently also cumbersone if you want to set a specific starting point for image to image with a fixed number of update steps afterwards, because you will most likely need to compromise on your exact starting point (or the number of timesteps).

**Describe the solution you'd like**
The easiest solution would be the option to overwrite the prev_timestep defined here:
https://github.com/huggingface/diffusers/blob/a6610db7a8ea0c8b04ec915b8ff3cac670606827/src/diffusers/schedulers/scheduling_ddim.py#L264
in the function signature of the step method.

**Describe alternatives you've considered**
An alternative would be to replace the prev_timestep calculation above using a function that can be passed to the scheduler. This seems slightly more complex for the user in the end, since there would be the need to actually define the the timestep function instead of just choosing any ordering of timesteps. On the plus side the user would not have to keep track of two timesteps at the same time.

**Additional context**
Could then be used like this:
```python
# create any timestep schedule here
timesteps = torch.linspace(1, 0, num_steps+1)
# change schedule shape
timesteps = timesteps ** 2
timesteps = (start_timestep * timesteps).long()
                                                                                                        
# add noise
noise = torch.randn_like(latents)
latents_noisy = scheduler.add_noise(latents, noise, timesteps[0])
for t, t_prev in zip(timesteps, timesteps[1:]):
    # pred noise
    latent_model_input = torch.cat([latents_noisy] * 2)
    latent_model_input = self.scheduler.scale_model_input(latent_model_input, t)
    noise_pred = self.unet(latent_model_input,
                           t,
                           encoder_hidden_states=text_embeddings,
                           ).sample
                                                                                                        
    noise_pred_uncond, noise_pred_text = noise_pred.chunk(2)
    noise_pred = noise_pred_uncond + guidance_scale * (noise_pred_text - noise_pred_uncond)
    latents_noisy = self.scheduler.step(noise_pred, t, latents_noisy, prev_timestep=t_prev).prev_sample
```

Here is an example using ten steps to create an image using stable diffusion, top row uses the linear schedule, bottom row the square schedule. Each column has the same initial noise. There is not too much difference between these two, but maybe there is some room for exploration.
![burger](https://user-images.githubusercontent.com/22528125/215510161-77090755-8ec8-46c8-b301-00cdd134f901.png)

If you would be interested in my preferred change, I am willing to provide a pull request, since I need this feature anyway.

**Comment by patrickvonplaten** [comment]
@jenuk, do you see any improvements when using a square scheduler instead of the linear one? 

Maybe we could introduce a new configuration variable called `schedule_type: str = {"linear", "square", ...}` ?

**Comment by jenuk** [comment]
I don't think I see drastic differences between the linear and square schedule right now, but I haven't compared them extensively. However, I just played a little around with the timesteps using only very few steps (3-5) and to me there seems to be a clear advantage (in using a non-linear schedule), even though all results obviously look worse than using many timesteps. Here are some examples: top row is a linear schedule with 50 steps, second row is a linear schedule with 3, 4, 5 steps, and last row is a non-liner schedule with 3, 4, 5 steps:
Prompt:  a photograph of a hamburger, highly detailed, 4k (3 steps)
![burger_few](https://user-images.githubusercontent.com/22528125/216657285-a46dd4e2-ace9-4239-9e51-eda23784df18.png)
Prompt: A robin (4 steps)
![bird_few](https://user-images.githubusercontent.com/22528125/216657292-49b9dfed-f669-4a2e-8d70-9e65b4d94e72.png)
Prompt: A tri-colored cat (5 steps)
![cat_few](https://user-images.githubusercontent.com/22528125/216657265-4eb49d4e-8dcd-4d42-869d-a2b8957456fd.png)

I can imagine that being useful for people that want to quickly find a desired initial noise.

Anyways, my initial thought was not that it necessarily enables better creations for stable diffusion specifically, but that the method itself allows flexibility that may be useful for specific goals (such as starting image-to-image at any timestep)  or maybe other models will just need a different schedule (such as the square schedule used  in the original paper for a single model) even if the current one works well for stable diffusion. And I don't see a reason to not include it.

Since my use case is actually concerned the flexibility itself, chosing from multiple fixed schedules would not be helpful for me, though I can see that this is much simpler solution for users in general.

**Comment by patrickvonplaten** [comment]
Also a somewhat related of how this issue could be solved: https://github.com/huggingface/diffusers/pull/2059/files#r1096093047
